drivers: entropy: Fix non-stop RNG ISR firing for STM32WB09

HoZHel opened the pull request:
- Fix the TRNG driver issue regarding non-stop ISR firing by clearing RNG_IRQ_SR_ERROR_IRQ flag.

- Exclude the WB0x series from discarding data 12 times from the RNG data register.

- Add a comment to the recover_seed_error function explaining the reason for reading and discarding data from the RNG data register 12 times.

- Fix the issue regarding passing the TRNG peripheral instance to the BLE driver.

- Increase the SYSTEM_WORKQUEUE_STACK_SIZE when CONFIG_BT is set.

Here is the screenshot of `beacon` sample to show the reason for increasing `SYSTEM_WORKQUEUE_STACK_SIZE`. Previously, the default was 1024 bytes.

<img width="971" height="434" alt="image" src="https://github.com/user-attachments/assets/d206c3ba-904e-4270-977b-8490795de995" />

HoZHel committed:
drivers: entropy: Fix non-stop RNG ISR firing for STM32WB09

Fix the TRNG driver issue regarding non-stop ISR firing for STM32WB09 by
clearing RNG_IRQ_SR_ERROR_IRQ flag.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
drivers: bluetooth: hci: Fix the issue about the TRNG peripheral instance

Fix the issue regarding passing the TRNG peripheral instance to the driver.

Increase the SYSTEM_WORKQUEUE_STACK_SIZE when CONFIG_BT is set.
According to the log of thread analyzer for beacon sample, 1048 bytes
are needed. So, it's been increased to a safer value.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

JarmouniA commented on drivers/bluetooth/hci/hci_stm32wb0.c:
```diff
@@ -343,7 +343,7 @@ static int ble_pm_action(const struct device *dev,
 static void rng_get_random(void *num, size_t size)
 {
-	const struct device *dev = DEVICE_DT_GET(DT_DRV_INST(0));
+	const struct device *dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
Relying on the chosen property here is unnecessary really, this is WB0 code, we know its RNG device, so we should get it by compatible using `DT_COMPAT_GET_ANY_STATUS_OKAY`.

etienne-lms reviewed (commented):
I don't understand how "drivers: entropy: Fix non-stop RNG ISR firing for STM32WB09" fixes your issue. I must miss something if it really changes the driver behavior.

For commit "drivers: bluetooth: hci: Fix the issue about the TRNG peripheral instance": could you describe a bit more why `CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE` is increased: how does it help? I also this the commit message header line and body is not very explicit about which TRNG issue is being addressed.

mathieuchopstm commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -327,10 +333,16 @@ static int recover_seed_error(RNG_TypeDef *rng)
 	}
 
+#if defined(CONFIG_SOC_STM32WB09XX)
+	if (LL_RNG_GetErrorIrq(rng)) {
```
\+1 on Etienne's comment: unless there's a reason for it, let's not bother checking for the flags.

HoZHel replied:
Do you mean clearing a flag without checking if it is raised?

HoZHel commented on drivers/entropy/entropy_stm32.h:
```diff
@@ -109,7 +109,9 @@ static inline rng_sample_t ll_rng_read_rand_data(RNG_TypeDef *RNGx)
 	 */
-	WRITE_REG(RNG->IRQ_SR, RNG_IRQ_SR_FF_FULL_IRQ);
+	if (LL_RNG_GetFfFullIrq(RNGx)) {
+		SET_BIT(RNGx->IRQ_SR, RNG_IRQ_SR_FF_FULL_IRQ);
```
What you're saying is the opposite. As the name suggests, `SET_BIT()` is for setting one bit of the register, while `WRITE_REG()` manipulates the whole register. [Here](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32wb0x/soc/stm32wb0x.h#L161-L171) you can find these two macros as follows:

`#define SET_BIT(REG, BIT)     ((REG) |= (BIT))`
`#define WRITE_REG(REG, VAL)   ((REG) = (VAL))`

mathieuchopstm replied:
I know what the macros do. `SET_BIT()` is not correct here because it does a read *then* write, and the flags are cleared by writing 1 to their bit:
<img width="655" height="448" alt="image" src="https://github.com/user-attachments/assets/7ff8a455-e78a-46ff-8e63-d758e6d687ff" />

If `ERROR_IRQ` was 1, `SET_BIT(IRQ_SR, FF_FULL_IRQ)` will read the register (`= ERROR_IRQ`), then `|= FF_FULL_IRQ` to the read value and write back the result (`= ERROR_IRQ | FF_FULL_IRQ`) which will clear both interrupt flags.

Using `WRITE_REG` (or really, just `RNG->IRQ_SR = RNG_IRQ_SR_FF_FULL_IRQ`) will write a value with only bit `FF_FULL_IRQ` set which ensures that only this specific interrupt flag is cleared.

etienne-lms commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -327,10 +333,16 @@ static int recover_seed_error(RNG_TypeDef *rng)
 
+#if defined(CONFIG_SOC_STM32WB09XX)
+	if (LL_RNG_GetErrorIrq(rng)) {
+		SET_BIT(rng->IRQ_SR, RNG_IRQ_SR_ERROR_IRQ);
```
My mistake. Indeed it is an error, we should clear that bit.
Could you rather fix `ll_rng_clear_sei()` adding their clear of IRQ_SR_ERROR?

I think we need to revisit a bit WB09xx sequences in this driver.

etienne-lms commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -36,6 +36,12 @@
+#if IRQLESS_TRNG
+#define MAX_AVAIL_RND_NUM 1
+#else
+#define MAX_AVAIL_RND_NUM 4 /* Internal four-word FIFO */
```
I don't think lowering the value will fix anything. Just a matter of consistency.

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -327,10 +333,16 @@ static int recover_seed_error(RNG_TypeDef *rng)
 
+#if defined(CONFIG_SOC_STM32WB09XX)
+	if (LL_RNG_GetErrorIrq(rng)) {
+		SET_BIT(rng->IRQ_SR, RNG_IRQ_SR_ERROR_IRQ);
```
Please use new bits handling functions: https://github.com/zephyrproject-rtos/zephyr/pull/97329

HoZHel commented:
> Could you mention that in the commit message? Like `Experimentation show that 1048 bytes of stack where used when BT is enabled hence increase CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE to a safer value.`.

Done.

etienne-lms reviewed (commented):
IMHO commit "drivers: entropy: Fix non-stop RNG ISR firing for STM32WB09" should only add clear ERROR_IRQ field of RNG[IR_SR]. Other changes should rather move to a dedicated commit. This could help backporting only this fix in v4.1 or v4.2 one needs to.

Commit "drivers: bluetooth: hci: Fix the issue about the TRNG peripheral instance" LGTM.

mathieuchopstm reviewed (changes requested):
Per remark of @etienne-lms, let's keep the previous implementation. Adding the comment is optional.

HoZHel committed:
drivers: entropy: Modify the recover_seed_error function for WB0x

Exclude the WB0x series from discarding data 12 times from the RNG data
register.

Add a comment to the recover_seed_error function explaining the reason
for reading and discarding data from the RNG data register 12 times.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel commented:
> IMHO commit "drivers: entropy: Fix non-stop RNG ISR firing for STM32WB09" should only add clear ERROR_IRQ field of RNG[IR_SR]. Other changes should rather move to a dedicated commit. This could help backporting only this fix in v4.1 or v4.2 one needs to.

Done.

mathieuchopstm reviewed (commented):
Otherwise LGTM.

etienne-lms commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -446,7 +451,7 @@ static uint16_t generate_from_isr(uint8_t *buf, uint16_t len)
 #if !IRQLESS_TRNG
 		NVIC_ClearPendingIRQ(IRQN);
-#endif /* IRQLESS_TRNG */
+#endif /* !IRQLESS_TRNG */
```
Could you mention this change in the commit log so that it does not appear as a unexpected change?

mathieuchopstm replied:
FTR GitHub won't invalidate approvals if you force-push but only commit messages have changed, if that's what bothers you

mathieuchopstm commented:
Adding backport to v4.1/v4.2 as TRNG driver support for WB09 was introduced in 4.1

mathieuchopstm commented:
added @etienne-lms as assignee as @erwango is out of office and can't approve

HoZHel commented:
> Adding backport to v4.1/v4.2 as TRNG driver support for WB09 was introduced in 4.1

@mathieuchopstm, I think we can't do backport since `stm32_reg_write()` is not present in [v4.2-branch](https://github.com/zephyrproject-rtos/zephyr/tree/v4.2-branch) and [v4.1-branch](https://github.com/zephyrproject-rtos/zephyr/tree/v4.1-branch).

etienne-lms commented:
That's true.
I've created an issue entry: https://github.com/zephyrproject-rtos/zephyr/issues/98668. Specific P-Rs for solving the issues on _v4.1-branch_  and _v4.2-branch_ will need to be created manually.
